src/contrib/mlir/compiler.cpp

```diff
@@ -0,0 +1,423 @@
+    static std::mutex mlir_init_mutex;
+    static bool initialized = false;
+
+    std::unique_lock<std::mutex> lock(mlir_init_mutex);
+
+    if (!initialized)
+    {
+        mlir::registerDialect<mlir::NGDialect>();
+        // Register any LLVM command line options
+        llvm::cl::ParseEnvironmentOptions("ngraph", "MLIR_LLVM_OPTIONS", "");
+        initialized = true;
+    }
+}
+
+void MLIRCompiler::compile_and_run()
+{
+    build_ng_dialect_module();
+    lower_ng_dialect();
+    optimize();
+    bind_arguments();
+    execute();
+    cleanup();
+}
+
+// Creates an MLIR module and function with nGraph dialect ops from the input CompiledKernel.
+void MLIRCompiler::build_ng_dialect_module()
+{
+    // initialize an empty module
+    m_module = make_unique<mlir::Module>(&m_context);
+
+    TypeList args_type_list, result_type_list;
+
+    // Retrieve input and output tensors.
+    const auto& kernel_inputs = m_compiled_kernel->get_arguments();
+    const auto& kernel_outputs = m_compiled_kernel->get_kernel_outputs();
+    NGRAPH_ASSERT(kernel_inputs.size() != 0) << "Cannot have empty inputs list";
+    NGRAPH_ASSERT(kernel_outputs.size() != 0) << "Cannot have empty outputs list";
+
+    for (auto input : kernel_inputs)
+    {
+        args_type_list.push_back(get_mlir_type(input->get_output_tensor_ptr().get()));
+    }
+
+    for (auto output : kernel_outputs)
+    {
+        result_type_list.push_back(get_mlir_type(output->get_output_tensor_ptr().get()));
+    }
+
+    auto func_type = mlir::FunctionType::get(args_type_list, result_type_list, &m_context);
+    auto function =
+        make_unique<mlir::Function>(mlir::UnknownLoc::get(&m_context), "main", func_type);
+    function->addEntryBlock();
+
+    // populate Tensor->Value maps
+    int i = 0;
+    for (auto input : kernel_inputs)
+    {
+        mlir::Value* arg = function->getArgument(i);
+        TensorInfo tensor_info{arg};
+        m_tensor_to_value_map.insert(
+            TensorToInfo(input->get_output_tensor_ptr().get(), tensor_info));
+        i++;
+    }
+
+    // create builder
+    m_builder = llvm::make_unique<mlir::FuncBuilder>(function.get());
+    build_ng_dialect();
+    m_module->getFunctions().push_back(function.release());
+    if (failed(m_module->verify()))
+    {
+        NGRAPH_FAIL() << "Invalid module after lowering to NG dialect";
+    }
+
+    dump_mlir_module("nGraph Dialect Dump:");
+}
+
+// Converts an nGraph Tensor into an MLIR tensor type, including the conversion of the Tensor's
+// element type.
+mlir::Type MLIRCompiler::get_mlir_type(const descriptor::Tensor* tensor)
+{
+    SmallVector<int64_t, 4> shape;
+    for (auto d : tensor->get_shape())
+    {
+        shape.push_back(d);
+    }
+
+    return mlir::NGTensorType::get(&m_context, get_mlir_type(tensor->get_element_type()), shape);
+}
+
+// Converts an nGraph element type into an MLIR type.
+mlir::Type MLIRCompiler::get_mlir_type(const element::Type& type)
+{
+    switch (type.get_type_enum())
```

Comment:
This should have the enum check like this
```
#if !(defined(__GNUC__) && (__GNUC__ == 4 && __GNUC_MINOR__ == 8))
#pragma GCC diagnostic push
#pragma GCC diagnostic error "-Wswitch"
#pragma GCC diagnostic error "-Wswitch-enum"
#endif

```
So that if a new datatype is added to ngraph then this will get a compile error until the new type is added to this switch

Reply:
Good suggestion. Changed